massa-protocol-worker-2/src/handlers/operation_handler/retrieval.rs

```diff
@@ -0,0 +1,471 @@
+                operations_prefix_ids: future_set,
+            });
+        }
+        if !ask_set.is_empty() {
+            {
+                let mut active_connections_write = self.active_connections.write();
+                if let Some(peer) = active_connections_write.connections.get_mut(peer_id) {
+                    peer.send_channels
+                        .send(
+                            &self.operation_message_serializer,
+                            OperationMessage::AskForOperations(ask_set).into(),
+                            false,
+                        )
+                        .map_err(|err| ProtocolError::SendError(err.to_string()))
+                } else {
+                    {
+                        let mut cache_write = self.cache.write();
+                        cache_write.ops_known_by_peer.pop(peer_id);
+                    }
+                    Ok(())
+                }
+            }
+        } else {
+            Ok(())
+        }
+    }
+
+    fn update_ask_operation(&mut self) -> Result<(), ProtocolError> {
+        let now = Instant::now();
+        while !self.op_batch_buffer.is_empty()
+        // This unwrap is ok because we checked that it's not empty just before.
+            && now >= self.op_batch_buffer.front().unwrap().instant
+        {
+            let op_batch_item = self.op_batch_buffer.pop_front().unwrap();
+            self.on_operations_announcements_received(
+                op_batch_item.operations_prefix_ids,
+                &op_batch_item.peer_id,
+            )?;
+        }
+        Ok(())
+    }
+
+    /// Maybe move this to propagation
+    /// Process the reception of a batch of asked operations, that means that
+    /// we have already sent a batch of ids in the network, notifying that we already
+    /// have those operations.
+    fn on_asked_operations_received(
+        &mut self,
+        peer_id: &PeerId,
+        op_pre_ids: OperationPrefixIds,
+    ) -> Result<(), ProtocolError> {
+        if op_pre_ids.is_empty() {
+            return Ok(());
+        }
+
+        let mut ops: Vec<SecureShareOperation> = Vec::with_capacity(op_pre_ids.len());
+        {
+            // Scope the lock because of the async call to `send_operations` below.
+            let stored_ops = self.storage.read_operations();
+            for prefix in op_pre_ids {
+                let opt_op = match stored_ops
+                    .get_operations_by_prefix(&prefix)
+                    .and_then(|ids| ids.iter().next())
+                {
+                    Some(id) => stored_ops.get(id),
+                    None => continue,
+                };
+                if let Some(op) = opt_op {
+                    ops.push(op.clone());
+                }
+            }
+        }
+        {
+            let mut active_connections_write = self.active_connections.write();
+            if let Some(peer) = active_connections_write.connections.get_mut(peer_id) {
```

Comment:
same, we can use `.get()` instead of `.get_mut()`

Reply:
Done